src/ui/Menus/PostponeMenu.ts

```diff
@@ -38,24 +38,24 @@ export class PostponeMenu extends TaskEditingMenu {
             itemNamingFunction: NamingFunction,
             postponingFunction: PostponingFunction,
         ) => {
+            const title = itemNamingFunction(task, amount, timeUnit);
+
+            item.setTitle(title).onClick(() =>
+                PostponeMenu.postponeOnClickCallback(button, task, amount, timeUnit, postponingFunction, taskSaver),
+            );
+
             // TODO some of the code below is duplicated in postponeOnClickCallback() and may be refactored
-            let isCurrentValue = false;
             const dateFieldToPostpone = getDateFieldToPostpone(task);
             if (dateFieldToPostpone) {
                 const { postponedDate } = postponingFunction(task, dateFieldToPostpone, timeUnit, amount);
 
                 if (task[dateFieldToPostpone]?.isSame(postponedDate, 'day')) {
-                    isCurrentValue = true;
+                    item.setChecked(true);
+                    // Workaround solution for https://github.com/obsidian-tasks-group/obsidian-tasks/issues/2816
+                    // Do nothing when the checked item is clicked
+                    item.onClick(() => null);
```

Comment:
What's the thinking for putting the code in here, rather than in `postponeOnClickCallback()`?

`postponeOnClickCallback()` is callable from multiple locations, and by putting the fix in the caller, it means all other callers also need to do the same work as this PR.

Reply:
1. We need to act on the `PostponeMenu`'s `item` which is not available in `PostponeMenu.postponeOnClickCallback()`. I can make it available there, but...
2. ...The other caller of `PostponeMenu.postponeOnClickCallback()` is the `QueryResultsRenderer` where it is used to process the click of postpone button, not to display the menu, so this functionality would be excessive there. Not to mention the absence of the `item` object in that call.

Reply:
Call in `QueryResultsRenderer` for reference:

```ts
button.addEventListener('click', (ev: MouseEvent) => {
            ev.preventDefault(); // suppress the default click behavior
            ev.stopPropagation(); // suppress further event propagation
            PostponeMenu.postponeOnClickCallback(button, task, amount, timeUnit);
        });
```

Reply:
No, this PR does not need to change the menu item.
It only needs to stop the menu item from being disabled if the date is unchanged.
That xan be done via early return in the call-back function if the date is unchanged.

Reply:
Right but I don't see the point of moving the fix to the callback. For now there are only 2 callers and the bug is related to one of the callers not the other one. So why the fix should apply to both?

Reply:
> Right but I don't see the point of moving the fix to the callback. For now there are only 2 callers and the bug is related to one of the callers not the other one. So why the fix should apply to both?

I believe that the cause of the bug is that the following line is reached when no edit has been made:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/dee3dd8eee0f418ced75d47994930c93594cdb11/src/ui/Menus/PostponeMenu.ts#L116

The closer the fix is to that location, the better.

As a general principle, it is cleaner and more sustainable to fix bugs at their location rather than to require the caller(s) to be modified... Off the top of my head, reasons include:

- If a new caller is added, the bug reoccurs, or if we are lucky, someone knows to duplicate the fix.
- Sometimes location of the bug gets fixed, and then we have redundant bug-fix code in the caller
- People modifying the location of the bug will not necessarily know about the bug-fix in the calling code...

Reply:
Ok, I understood you point of view and modified the PR, thanks a lot for the details.

I'm thinking now that the line you mentioned might even be unnecessary. What do you think? I tried postponing to "checked" dates (postpone to current date) and normal dates (real postpone) and there seems to be no issue.

Or maybe the fix should be the early return (as in this PR for now) + the removal of the code you mentioned. What do you think?

As per my understanding of async/await in JS (I double confirmed [here](https://chatgpt.com/share/66e3e142-6484-8007-b960-4e27be8bab6b)), all the update operations shall be completed before the `PostponeMenu.postponeSuccessCallback()` is called.

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/dee3dd8eee0f418ced75d47994930c93594cdb11/src/ui/Menus/PostponeMenu.ts#L106-L107

Also this is the only reference to `button` in `PostponeMenu`, so removing it would remove a lot of code.

Reply:
Thanks.

> when the task date to postpone and the postponed date in the menu match, do nothing (remove the previously assigned item click handler)

Trivial: In the PR description, I think the bit in `()` can be removed now...

I will review in test early next week...

Reply:
Is the new code location already covered by tests? If so, could  this be covered by checking that the same Task object is returned? (I haven't checked the code to know if this is possible - just a thought....)

Reply:
> Trivial: In the PR description, I think the bit in `()` can be removed now...

oh, thanks, I changed the description.

> Is the new code location already covered by tests? If so, could this be covered by checking that the same Task object is returned? (I haven't checked the code to know if this is possible - just a thought....)

There are no tests on `PostponeMenu.postponeOnClickCallback()`.

Reply:
> I'm thinking now that the line you mentioned might even be unnecessary. What do you think? I tried postponing to "checked" dates (postpone to current date) and normal dates (real postpone) and there seems to be no issue.

I think I did some experimenting and found that when I removed it, occasionaly later edits failed...

Reply:
Resolving conversation so I can merge...